game-core/src/system/ally.rs

```diff
@@ -1,81 +1,131 @@
-    }
-}
-
-pub struct Grouper;
-
-impl<'s> System<'s> for Grouper {
     type SystemData = (
         ReadStorage<'s, Ally>,
-        WriteStorage<'s, Transform>,
-        WriteStorage<'s, Parent>,
-        WriteStorage<'s, Player>,
+        WriteStorage<'s, Motion>,
+        ReadStorage<'s, Transform>,
+        ReadStorage<'s, Player>,
+        Read<'s, GameoffConfig>,
         Entities<'s>,
     );
 
     fn run(
         &mut self,
-        (allies, mut transforms, mut parents, mut players, entities): Self::SystemData,
+        (allies, mut motions, transforms, players, config, entities): Self::SystemData,
     ) {
-        let get_relative_position = |mut num_allies: u32| {
-            num_allies %= 9;
+        let mut rng = rand::thread_rng();
+        let zero_distance_dist = Uniform::new(0.5, 1.0);
 
-            let ind = if num_allies == 4 {
-                num_allies + 1 // skip player position
-            } else {
-                num_allies
-            } as i32;
+        let p_transform = {
+            let (t, _) = (&transforms, &players)
+                .join()
+                .next()
+                .expect("no player found");
+            (t.clone())
+        };
 
-            let row = ind / 3 - 1;
-            let col = ind % 3 - 1;
+        for (_, motion, transform1, entity1) in
+            (&allies, &mut motions, &transforms, &entities).join()
+        {
+            let d = (p_transform.translation - transform1.translation).truncate();
+            let m = d.magnitude().abs();
+
+            let mut pv = d.normalize(); //unit vector
+                                        //less than follow distance, do nothing
+            if m < config.ally.follow_distance {
+                pv *= 0.0;
+            } else if (m > config.ally.follow_distance) && (m < config.ally.max_distance) {
+                //at or more than the follow distance, play a bit of catchup
+                pv *= (1.0
+                    + (m - config.ally.follow_distance)
+                        / (config.ally.max_distance - config.ally.follow_distance))
+                    * config.speed;
+            } else if m > config.ally.max_distance {
+                //max catchup is 2.0
+                pv *= 2.0 * config.speed;
+            }
 
-            (row as f32 * 32.0, col as f32 * 32.0)
-        };
+            let mut av = Vector2::new(0.0, 0.0);
+
+            for (_, transform2, entity2) in (&allies, &transforms, &entities).join() {
+                if entity1 != entity2 {
+                    let d = (transform1.translation - transform2.translation).truncate();
+                    let m = d.magnitude().abs();
+                    let mut v = if m == 0.0 {
+                        //make sure the vector always points somewhere
+                        Vector2::new(
+                            zero_distance_dist.sample(&mut rng),
+                            zero_distance_dist.sample(&mut rng),
```

Comment:
I don't understand whats happening here. It looks like you're getting a random vector of {0.5-1.0, 0.5-1.0}. Is this just to force it to move somewhere if they're right on top of each other? I mostly just find it odd that its pretty much always going to be pointing up and to the right. It might make more sense to rand for a number of degrees and then use some trig to generate a unit vector in that direction. This is fine though, so I don't mind merging it as is. Just an idea.